### PR TITLE
Small sam patch

### LINKED ISSUE #22904: SAM: Notebook example not working
### System Info

- `transformers` version: 4.29.0.dev0
- Platform: macOS-13.2-arm64-arm-64bit
- Python version: 3.10.6
- Huggingface_hub version: 0.13.4
- Safetensors version: 0.3.0
- PyTorch version (GPU?): 1.13.0 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): 0.6.9 (cpu)
- Jax version: 0.4.8
- JaxLib version: 0.4.7
- Using GPU in script?: NO
- Using distributed or parallel set-up in script?: NO

Dependencies
- torch = 1.13.0
- numpy = 1.23.4

### Who can help?

_No response_

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Pull [SAM Notebook example](https://github.com/huggingface/notebooks/blob/main/examples/segment_anything.ipynb)
2. Run notebook up until 
```
masks = processor.image_processor.post_process_masks(outputs.pred_masks.cpu(), inputs["original_sizes"].cpu(), inputs["reshaped_input_sizes"].cpu())
```
3. Get error
```
TypeError: upsample_bilinear2d() received an invalid combination of arguments - got (Tensor, list, bool, NoneType), but expected one of:
 * (Tensor input, tuple of SymInts output_size, bool align_corners, tuple of floats scale_factors)
      didn't match because some of the arguments have invalid types: (Tensor, !list!, bool, !NoneType!)
 * (Tensor input, tuple of SymInts output_size, bool align_corners, float scales_h, float scales_w, *, Tensor out)
```

### Expected behavior

original_sizes/output_sizes to be of the expected type, is this a dependency issue?

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Fixes #22904 . 
It is backward compatible and prevents having to modify any of the notebooks we shared

### CODE REVIEW

File: tests/models/sam/test_modeling_sam.py [hunk 1]
```diff
@@ -733,3 +733,20 @@ def test_inference_mask_generation_three_boxes_point_batch(self):
         iou_scores = outputs.iou_scores.cpu()
         self.assertTrue(iou_scores.shape == (1, 3, 3))
         torch.testing.assert_allclose(iou_scores, EXPECTED_IOU, atol=1e-4, rtol=1e-4)
+
+    def test_post_process_masks(self):
```
Comment (amyeroberts) [inline tests/models/sam/test_modeling_sam.py:737]: This isn't a model test - it should be with the processing tests

Comment (ArthurZucker) [reply tests/models/sam/test_modeling_sam.py:737]: Yes and no, it need the outputs from a model

Comment (ArthurZucker) [reply tests/models/sam/test_modeling_sam.py:737]: I could! Just wanted to make sure that the post processing performs what we want it to perform.
Will add dummy masks.

Comment (ArthurZucker) [reply tests/models/sam/test_modeling_sam.py:737]: Also it was good to test the `reshaped_input_sizes` that were in tensor format (from being outputed by the model).
So I still wanna make sure that the post process can run from the outputs of the model.

**Review by sgugger — APPROVED** [review]
LGTM but will defer to Amy who knows the image processor better than me :-)

File: src/transformers/models/sam/image_processing_sam.py [hunk 2]
```diff
@@ -398,9 +399,16 @@ def post_process_masks(
...
+        if isinstance(original_sizes, (torch.Tensor, np.ndarray)):
+            original_sizes = original_sizes.tolist()
+        if isinstance(reshaped_input_sizes, (torch.Tensor, np.ndarray)):
+            reshaped_input_sizes = reshaped_input_sizes.tolist()
         output_masks = []
         for i, original_size in enumerate(original_sizes):
+            if isinstance(masks[i], np.ndarray):
+                masks[i] = torch.from_numpy(masks[i])
+            elif isinstance(masks[i], List):
+                raise ValueError("Input masks should be a list of `torch.tensors` or a list of `np.ndarray`")
```
Comment (amyeroberts) [inline src/transformers/models/sam/image_processing_sam.py:411]: I would just have `else` here - testing for list is very specific (is there a reason for checking this specific type?) 
```suggestion
            else:
                raise ValueError("Input masks should be a list of `torch.tensors` or a list of `np.ndarray`")
```

File: src/transformers/models/sam/image_processing_sam.py [hunk 3]
```diff
@@ -378,12 +378,13 @@ def post_process_masks(
...
         Args:
-            masks (`torch.Tensor`):
+            masks (`Union[List[torch.Tensor], List[np.ndarray]]`):
                 Batched masks from the mask_decoder in (batch_size, num_channels, height, width) format.
-            original_sizes (`torch.Tensor`):
-                The original size of the images before resizing for input to the model, in (height, width) format.
-            reshaped_input_sizes (`torch.Tensor`):
-                The size of the image input to the model, in (height, width) format. Used to remove padding.
+            original_sizes (`Union[torch.Tensor, List[Tuple[int,int]]]`):
+                The original sizes of each image before it was resize to the model's expected input shape, in (height,
```
Comment (amyeroberts) [inline src/transformers/models/sam/image_processing_sam.py:384]:
```suggestion
                The original sizes of each image before it was resized to the model's expected input shape, in (height,
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and iterating!

File: src/transformers/models/sam/image_processing_sam.py (hunk 2 above)
Comment (younesbelkada) [reply src/transformers/models/sam/image_processing_sam.py]: Hmm I think this seems to be needed, I just added another check for testing if `masks[i]` is a tensor

Comment (younesbelkada) [reply src/transformers/models/sam/image_processing_sam.py]: Otherwise the post processing fails

**Pull request merged**